libraries/AP_Compass/AP_Compass_BMM150.cpp

```diff
@@ -0,0 +1,279 @@
+    _dig_z3 = (int16_t)le16toh(data[DIG_Z3_LSB_REG - DIG_X1_REG] | data[DIG_Z3_MSB_REG - DIG_X1_REG] << 8);
+    _dig_z4 = (int16_t)le16toh(data[DIG_Z4_LSB_REG - DIG_X1_REG] | data[DIG_Z4_MSB_REG - DIG_X1_REG] << 8);
+    _dig_xyz1 = (uint16_t)le16toh(data[DIG_XYZ1_LSB_REG - DIG_X1_REG] | data[DIG_XYZ1_MSB_REG - DIG_X1_REG] << 8);
+}
+
+bool AP_Compass_BMM150::init()
+{
+    hal.scheduler->suspend_timer_procs();
+
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        hal.console->printf("BMM150: Unable to get bus semaphore\n");
+        goto fail_sem;
+    }
+
+    /* Do a soft reset */
+    _dev->write_register(POWER_AND_OPERATIONS_REG, SOFT_RESET);
+    hal.scheduler->delay(2);
+
+    /* Change power state from suspend mode to sleep mode */
+    _dev->write_register(POWER_AND_OPERATIONS_REG, POWER_CONTROL_VAL);
+    hal.scheduler->delay(2);
+
+    uint8_t val;
+    _dev->read_registers(CHIP_ID_REG, &val, 1);
+    if (val != CHIP_ID_VAL) {
+        hal.console->printf("BMM150: Wrong id\n");
+        goto fail;
+    }
+
+    _load_trim_values();
+
+    /*
+     * Recommended preset for high accuracy:
+     * - Rep X/Y = 47
+     * - Rep Z = 83
+     * - ODR = 20
+     * But we are going to use 30Hz of ODR
+     */
+    _dev->write_register(REPETITIONS_XY_REG, (47 - 1) / 2);
+    _dev->write_register(REPETITIONS_Z_REG, 83 - 1);
+    /* Change operation mode from sleep to normal and set ODR */
+    _dev->write_register(OP_MODE_SELF_TEST_ODR_REG, NORMAL_MODE | ODR_30HZ);
+
+    _dev->get_semaphore()->give();
+    hal.scheduler->resume_timer_procs();
+
+    /* register the compass instance in the frontend */
+    _compass_instance = register_compass();
+    set_dev_id(_compass_instance, AP_COMPASS_TYPE_BMM150);
+
+    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&AP_Compass_BMM150::_update, void));
+
+    return true;
+
+fail:
+    _dev->get_semaphore()->give();
+fail_sem:
+    hal.scheduler->resume_timer_procs();
+    return false;
+}
+
+/*
+ * Compensation algorithm got from https://github.com/BoschSensortec/BMM050_driver
+ * this is not explained in datasheet.
+ */
+int16_t AP_Compass_BMM150::_compensate_xy(int16_t xy, uint16_t rhall, bool is_x)
+{
+    int8_t txy1, txy2;
+    int32_t val;
+
+    if (!rhall) {
+        rhall = _dig_xyz1;
+    }
+
+    if (is_x) {
+        txy1 = _dig_x1;
+        txy2 = _dig_x2;
+    } else {
+        txy1 = _dig_y1;
+        txy2 = _dig_y2;
+    }
+
+    /* TODO: make compensation calculations works on big-endian CPUs */
+    val = ((int16_t)(((uint16_t)((((int32_t)_dig_xyz1) << 14) / rhall)) -
+          ((uint16_t)0x4000)));
+    val = ((int16_t)((((int32_t)xy) * ((((((((int32_t)_dig_xy2) *
+          ((((int32_t)val) * ((int32_t)val)) >> 7)) + (((int32_t)val) *
+          ((int32_t)(((int16_t)_dig_xy1) << 7)))) >> 9) +
+          ((int32_t)0x100000)) * ((int32_t)(((int16_t)txy2) +
```

Comment:
Wow. -ETOOMANYCASTS. if you reduce the number of casts, maybe it's possible to remove the TODO above